src/parser.ts

```diff
@@ -0,0 +1,177 @@
+ */
+
+import { model, Raml10Resolver } from "amf-client-js";
+import amf from "amf-client-js";
+import path from "path";
+import _ from "lodash";
+
+import { FatRamlResourceLoader } from "./exchange-connector";
+import { RestApi } from ".";
+
+/**
+ * Parses a RAML file to an AMF model.
+ *
+ * @param filename The path to the RAML file to load
+ *
+ * @returns an AMF model of the RAML
+ */
+export async function parseRamlFile(
+  filename: string
+): Promise<amf.model.document.BaseUnit> {
+  const fileUri = `file://${path.resolve(filename)}`;
+
+  // We initialize AMF first
+  amf.plugins.document.WebApi.register();
+  amf.plugins.document.Vocabularies.register();
+  amf.plugins.features.AMFValidation.register();
+
+  await amf.Core.init();
+
+  const fatRamlResourceLoader = new FatRamlResourceLoader(
+    path.dirname(fileUri)
+  );
+  const ccEnvironment = amf.client.DefaultEnvironment.apply().addClientLoader(
+    fatRamlResourceLoader
+  );
+  const parser = new amf.Raml10Parser(ccEnvironment);
+
+  let model: amf.model.document.BaseUnit;
+  try {
+    model = await parser.parseFileAsync(fileUri);
+  } catch (err) {
+    console.log("Error validating");
+    console.log(err);
+  }
+  return model;
+}
+
+function getDataTypesFromDeclare(
+  types: model.domain.DomainElement[],
+  existingDataTypes: Set<string>
+): model.domain.CustomDomainProperty[] {
+  const ret: model.domain.CustomDomainProperty[] = [];
+  types.forEach((dataType: model.domain.CustomDomainProperty) => {
+    if (
+      !existingDataTypes.has(dataType.name.value()) &&
+      dataType.name.value() !== "trait"
+    ) {
+      existingDataTypes.add(dataType.name.value());
+      ret.push(dataType);
+    }
+  });
+  return ret;
+}
+
+/**
+ * Get all the referenced data types
+ *
+ * @param apiReferences - Array of references
+ * @param dataTypes - Array of data types
+ * @param existingDataTypes - Set of names of data types, used to de-duplicate the data types
+ */
+export function getReferenceDataTypes(
+  apiReferences: model.document.BaseUnit[],
+  dataTypes: model.domain.CustomDomainProperty[],
+  existingDataTypes: Set<string>
+): void {
+  if (apiReferences == null || apiReferences.length == 0) {
+    return;
+  }
+  apiReferences.forEach(
+    (reference: model.document.BaseUnit & amf.model.document.DeclaresModel) => {
+      if (reference.declares) {
+        dataTypes.push(
+          ...getDataTypesFromDeclare(reference.declares, existingDataTypes)
+        );
+      }
+      getReferenceDataTypes(
+        reference.references(),
+        dataTypes,
+        existingDataTypes
+      );
+    }
+  );
+}
+
+/**
+ * Extract all of the delcared data types from an AMF model.
+ *
+ * @param api - The model to extract data types from
+ *
+ * @returns data types from model
+ */
+export function getAllDataTypes(
+  api: model.document.BaseUnit & amf.model.document.DeclaresModel
```

Comment:
This might be my fault, but why `model.document..` and `amf.model.document...`

Reply:
Fixed